### ISSUE TITLE
About the file name

### ISSUE THREAD

**Issue by hanjun-lin** [tus/tusd#44]
Hi all,

I tried to run the official binary distribution of Go language edition of tud with official Android sample code (client). I found a file name problem (or may be an enhancement / feature request) of current sample codes of server side and client side.

The uploaded file's file name on the server is convert to a **random hashed file name** (e.g. _5dcecfd73c6d77bf7dfcf4fbdd75275e.bin_ with _5dcecfd73c6d77bf7dfcf4fbdd75275e.info_ )
It's a good design to avoid the duplicate of file name of uploaded files.
But the server side do not know the original file name of uploaded file either in _.info_ file. (_.info_ file do not have the information of original file name)
The client side is also do not know the new file name of uploaded file.
So it is difficult to make the relationship of uploaded file name with original file name for further application (e.g. client side can call the server API to tell the server the uploaded file name for server to fetch the file for further processing after client side uploaded the file)
The file extensions (e.g. _.png_, _.zip_) are also changed to .bin. The server side is also difficult to know the original file extension name to do further use.

In conclusion, my suggestions are as below:

1. **Server side** (tud):
    a. change the file name to the format such like the "**original_file_name_random_hash.original_file_extension**" (e.g. _MyPicture_e4fe1bebdfcdc5efd402de1d2bfa6ac4.png_)
    b. record the original file name information and new file name information in the _.info_ file for further processing
    c. return the new file name to the client side via the protocol or some mechanism

2. **Client side** (sample codes)
     a. receive and show the "new file name" information of uploaded file

**Comment by Acconut** [comment]
The reason why the tusd server generates random IDs for uploads instead of using the file's name is that the protocol (see http://tus.io) does not force the client to tell the server the file name. This is a decision that has been made on purpose. However, we do understand that such information is sometimes crucial and therefore you are able to use the mechanics of meta data to pass it along. In this case the file name will be presented as meta data inside the `.info` object.

To add the file name using our Android/Java client, you have to use the `TusUpload#setMetadata` method and pass it a map containing the information. Right now, this is not automatically done but I will enable to by default ASAP.

**Comment by Acconut** [comment]
> Right now, this is not automatically done but I will enable to by default ASAP.

The code for this change is already in the Android and Java repositories and will be released very soon.

**Comment by hanjun-lin** [comment]
Thanks a lot. Latest commits of client sides are work now.

**Comment by Acconut** [comment]
> But the performance and efficiency is not good for production because of the file I/O operating for reading the .info file.

Please explain this thought more. You need a place to save the additional information about the upload and this is exactly why we need it. We cannot simply remove it. Also, if you have bigger uploads, the impact of reading this small file is negligible IMO.

> It will be good if the file extension name stored in the server is same as uploaded file. (e.g. uploaded file extension name is .png and server stored file extension name is .png [currently is .bin])

As I said, we must not assume that we have a file name available and therefore cannot use the "original" extension. In addition, this would make the file handling logic more complex without providing much benefit to tusd or am I missing something?

**Comment by hanjun-lin** [comment]
> > It will be good if the file extension name stored in the server is same as uploaded file. (e.g. uploaded file extension name is .png and server stored file extension name is .png [currently is .bin])

> As I said, we must not assume that we have a file name available and therefore cannot use the "original" extension. In addition, this would make the file handling logic more complex without providing much benefit to tusd or am I missing something?

I agree with you. It may a good viewpoint for a protocol and sample code of keeping the file handing logic simple. The file handling logic can be implemented by another agent or API written by whom wants to do further applications.


> > But the performance and efficiency is not good for production because of the file I/O operating for reading the .info file.

> Please explain this thought more. You need a place to save the additional information about the upload and this is exactly why we need it. We cannot simply remove it. Also, if you have bigger uploads, the impact of reading this small file is negligible IMO.

It's not my mean to remove the .info file. In my original simple thoughts, the original file name can be as full or partial of new file name in the server. It will be more efficient without reading the .info file to get the original file name in the server side. It just a simple thoughts. Current implementations and sample codes I think is enough for further applications.

**Comment by Acconut** [comment]
> I agree with you.

:+1: 

> It's not my mean to remove the .info file. In my original simple thoughts, the original file name can be as full or partial of new file name in the server. It will be more efficient without reading the .info file to get the original file name in the server side. It just a simple thoughts.

That's a good use case but a very specialized one. In general, I think, people will need more information than just the original file name and therefore have to read the additional data anyway. So I am sorry, but your suggestion won't make it into tusd.
